component/src/main/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestAuthenticator.java

```diff
@@ -0,0 +1,396 @@
+			String tokenEndPoint = getTokenEndpoint(authenticatorProperties);
+			String callbackUrl = getCallbackUrl(authenticatorProperties);
+			OAuthAuthzResponse authzResponse = OAuthAuthzResponse.oauthCodeAuthzResponse(request);
+			String code = authzResponse.getCode();
+			OAuthClientRequest accessRequest;
+			try {
+				accessRequest =
+						OAuthClientRequest.tokenLocation(tokenEndPoint).setGrantType(GrantType.AUTHORIZATION_CODE)
+						                  .setClientId(clientId).setClientSecret(clientSecret)
+						                  .setRedirectURI(callbackUrl).setCode(code).buildBodyMessage();
+				// create OAuth client that uses custom http client under the hood
+				OAuthClient oAuthClient = new OAuthClient(new URLConnectionClient());
+				OAuthClientResponse oAuthResponse;
+				oAuthResponse = oAuthClient.accessToken(accessRequest);
+				String accessToken = oAuthResponse.getParam(PinterestAuthenticatorConstants.ACCESS_TOKEN);
+				if (StringUtils.isNotEmpty(accessToken)) {
+					Map<ClaimMapping, String> claims = buildClaims(oAuthResponse, authenticatorProperties);
+					if (claims != null && !claims.isEmpty()) {
+						//Find the subject from the IDP claim mapping, subject Claim URI.
+						String subjectFromClaims = FrameworkUtils
+								.getFederatedSubjectFromClaims(context.getExternalIdP().getIdentityProvider(), claims);
+						setSubjectFromClaims(context, subjectFromClaims, claims);
+					} else {
+						throw new AuthenticationFailedException("Selected user profile not found");
+					}
+				} else {
+					throw new AuthenticationFailedException("Authentication Failed, access token not available");
+				}
+			} catch (OAuthSystemException e) {
+				throw new AuthenticationFailedException(
+						"Exception while building a request for requesting the access token", e);
+			} catch (ApplicationAuthenticatorException e) {
+				throw new AuthenticationFailedException("Error while building the claim mapping", e);
+			}
+		} catch (OAuthProblemException e) {
+			throw new AuthenticationFailedException("Authentication Failed in OAuthClientResponse", e);
+		}
+	}
+
+	/**
+	 * Extra request sending to Pinterest userinfo end-point.
+	 *
+	 * @param url         user info endpoint.
+	 * @param accessToken access token.
+	 * @return user info
+	 * @throws ApplicationAuthenticatorException
+	 */
+	private String getUserInfo(String url, String accessToken) throws ApplicationAuthenticatorException {
+		if (log.isDebugEnabled()) {
+			log.debug("Sending the request for getting the user info");
+		}
+		StringBuilder stringBuilder = new StringBuilder();
+		BufferedReader bufferedReader = null;
+		HttpURLConnection httpConnection = null;
+		try {
+			URL obj = new URL(url + PinterestAuthenticatorConstants.QUESTION_MARK +
+			                  PinterestAuthenticatorConstants.ACCESS_TOKEN + PinterestAuthenticatorConstants.EQUAL +
+			                  accessToken);
+			URLConnection connection = obj.openConnection();
```

Comment:
Better to set some timeout ,
e.g.
connection.setConnectTimeout(connectTimeout);
connection.setReadTimeout(socketTimeout);